### PR TITLE
WIP: fix for https://github.com/aviatesk/JET.jl/issues/277 

### PR DESCRIPTION
**Pull request by goerch**
Unfortunately some `eval` tests are broken. I currently have no idea how to fix those...

It seems we have to adapt `select_statements` to `JuliaInterpreter`'s frame then?

### CODE REVIEW

File: src/toplevel/virtualprocess.jl [hunk 1]
```diff
@@ -796,7 +796,11 @@ function partially_interpret!(interp::ConcreteInterpreter, mod::Module, src::Cod
         print_with_code(io, src, concretize)
     end
 
-    selective_eval_fromstart!(interp, Frame(mod, src), concretize, #= istoplevel =# true)
+    # Frame src might differ from JET src?
+    framecode = JuliaInterpreter.FrameCode(mod, src, optimize=false)
+    @assert length(framecode.src.code) == length(concretize)
+    frame = Frame(framecode, JuliaInterpreter.prepare_framedata(framecode, []))
```
Comment (aviatesk) [inline src/toplevel/virtualprocess.jl:802]:
```suggestion
    # NOTE if `JuliaInterpreter.optimize!` may modify `src`, `src` and `concretize` can be inconsistent
    # here we create `JuliaInterpreter.Frame` by ourselves disabling the optimization (#277)
    framecode = JuliaInterpreter.FrameCode(mod, src, optimize=false)
    @assert length(framecode.src.code) == length(concretize)
    frame = Frame(framecode, JuliaInterpreter.prepare_framedata(framecode, Any[]))
```

**Review by aviatesk — COMMENTED** [review]
Good catch! Are you willing to add a testcase for this?
In a case of when `JuliaInterpreter.optimize!` is _necessary_ for successful interpretation, 
then as an alternative, we may even change `JuliaInterpreter.optimize!` so that it takes a callback that is invoked upon source modification.
This is just a suggestion, if you feel it would be complicated, we can just go ahead and merge this.

@goerch Also, here're tips for you when working on git projects and GitHub :)
- in order to keep commit history clean, it's generally better to keep the default branch of your fork up-to-dated with that of the original repository (i.e. `git pull origin (master|main)` occasionally), and cut a separate branch when working on something and create a PR from there
- in commit messages, it would give us extra fancy features if you use issue numbers (e.g. `#277`) rather than entire URL (e.g. `https://github.com/aviatesk/JET.jl/issues/277`), like GitHub automatically closes the referenced issue upon merging this PR

**Comment by goerch** [comment]
> Good catch! Are you willing to add a testcase for this?

Yes.

> In a case of when `JuliaInterpreter.optimize!` is _necessary_ for successful interpretation,
> then as an alternative, we may even change `JuliaInterpreter.optimize!` so that it takes a callback that is invoked upon source modification.

This version of JET currently produces 

```
═════ 1 toplevel error found ═════
┌ @ C:\Users\Win10\.julia\packages\ScanByte\BvBXC\src\codegen.jl:20 
│ invalid lookup expr Core.svec()
│ Stacktrace:
│   [1] error(::String, ::Expr)
│     @ Base .\error.jl:42
│   [2] lookup_expr(frame::JuliaInterpreter.Frame, e::Expr)
│     @ JuliaInterpreter C:\Users\Win10\.julia\packages\JuliaInterpreter\4B89D\src\interpret.jl:25
│   [3] collect_args(frame::JuliaInterpreter.Frame, call_expr::Expr; isfc::Bool)
│     @ JuliaInterpreter C:\Users\Win10\.julia\packages\JuliaInterpreter\4B89D\src\interpret.jl:60
│   [4] evaluate_foreigncall(frame::JuliaInterpreter.Frame, call_expr::Expr)
│     @ JuliaInterpreter C:\Users\Win10\.julia\packages\JuliaInterpreter\4B89D\src\interpret.jl:151
│   [5] eval_rhs(recurse::Any, frame::JuliaInterpreter.Frame, node::Expr)
│     @ JuliaInterpreter C:\Users\Win10\.julia\packages\JuliaInterpreter\4B89D\src\interpret.jl:402
│   [6] step_expr!(recurse::Any, frame::JuliaInterpreter.Frame, node::Any, istoplevel::Bool)
│     @ JuliaInterpreter C:\Users\Win10\.julia\packages\JuliaInterpreter\4B89D\src\interpret.jl:474
│   [7] step_expr!(interp::JET.ConcreteInterpreter{JET.JETAnalyzer{JET.BasicPass{typeof(JET.basic_function_filter)}}, Expr}, frame::JuliaInterpreter.Frame, node::Any, istoplevel::Bool)
│     @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\toplevel\virtualprocess.jl:974
│   [8] step_expr!(recurse::Any, frame::JuliaInterpreter.Frame, istoplevel::Bool)
│     @ JuliaInterpreter C:\Users\Win10\.julia\packages\JuliaInterpreter\4B89D\src\interpret.jl:611
│   [9] selective_eval!(recurse::Any, frame::JuliaInterpreter.Frame, isrequired::Vector{Bool}, istoplevel::Bool)
│     @ LoweredCodeUtils C:\Users\Win10\.julia\packages\LoweredCodeUtils\jZY56\src\codeedges.jl:850
│  [10] selective_eval_fromstart!
│     @ C:\Users\Win10\.julia\packages\LoweredCodeUtils\jZY56\src\codeedges.jl:875 [inlined]
│  [11] partially_interpret!(interp::JET.ConcreteInterpreter{JET.JETAnalyzer{JET.BasicPass{typeof(JET.basic_function_filter)}}, Expr}, mod::Module, src::Core.CodeInfo)
│     @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\toplevel\virtualprocess.jl:807
│  [12] _virtual_process!(toplevelex::Expr, filename::String, analyzer::JET.JETAnalyzer{JET.BasicPass{typeof(JET.basic_function_filter)}}, config::JET.ToplevelConfig{Expr}, context::Module, res::JET.VirtualProcessResult)
│     @ JET C:\Users\Win10\Documents\GitHub\JET.jl\src\toplevel\virtualprocess.jl:630
└───────────────────────────────────────────────────────────────────
```

for #224, so it seems to me we somehow depend on optimized code? Do you have an idea where I could search for the cause of the above?

**Comment by aviatesk** [comment]
Maybe https://github.com/JuliaDebug/JuliaInterpreter.jl/issues/13#issuecomment-464880123?
Or we can create `Frame` with optimization first and then select statements.

**Comment by goerch** [comment]
> Or we can create `Frame` with optimization first and then select statements.

This I tried first :) and it had too much fallout in JET for me to handle.

> Maybe JuliaDebug/JuliaInterpreter.jl#13 (comment)?

I will study this now. Thanks!

**Comment by goerch** [comment]
I'd like to propose the following:

- Merging this change as is (it seems to be a bug fix at least)
- Me learning some git (Thanks for the tips!)
- Switching internally to optimized frames in another pull request and talking there about the `eval` problems.

**Comment by aviatesk** [comment]
Sure!

**Pull request merged**